### PR TITLE
Auto refresh imenu

### PR DESCRIPTION
**Pull request by jcs090218**
This PR is for feature #511 and #295. 

I have currently bind to only `after-save-hook`. Open for any other suggestion! Thanks!

### CODE REVIEW

**Comment by brotzeit** [comment]
I think `after-save-hook` is fine for now, but @yyoncho suggested using a timer. IMO we should merge this and update the code when we have a more sophisticated solution.

**Comment by jcs090218** [comment]
> I think `after-save-hook` is fine for now, but @yyoncho suggested using a timer. IMO we should merge this and update the code when we have a more sophisticated solution.

I was thinking about using a timer too. This patch is indeed a quick solution; I am looking forward to update this in the near future.

File: lsp-ui-imenu.el
Comment (brotzeit) [inline lsp-ui-imenu.el:341]: I wonder if we should make this function a command and bind it to "r" in `lsp-ui-imenu-mode`.

Comment (jcs090218) [reply lsp-ui-imenu.el:342]: Added!

**Comment by jcs090218** [comment]
I have managed to use `timer` and `after-change-functions`. I also added `lsp-ui-util.el` file, but currently there aren't much yet. What do guys think about the `utility` module? I think is quite nice to have a `helper` file that holds all the generic functions/macros. Thanks!

**Comment by brotzeit** [comment]
Tested it and seems to work correctly :+1:

**Comment by brotzeit** [comment]
@jcs090218 do you know that you can close issues automatically https://docs.github.com/en/enterprise/2.13/user/articles/closing-issues-using-keywords ?

**Comment by jcs090218** [comment]
> do you know that you can close issues automatically

Yeah, I think I knew it, but I like to do it separately. Maybe I am just not use to this feature yet.

@brotzeit I guess we can merge this now and improve in the next patch?

Thanks!

**Comment by brotzeit** [comment]
I'd like to hear @sebastiencs opinion first.

File: README.md
Comment (brotzeit) [inline README.md:112]: `lsp-ui-imenu-refresh-delay` is missing

File: README.md
Comment (brotzeit) [inline README.md:111]: menu -> imenu

Comment (jcs090218) [reply README.md]: Fixed!

File: README.md
Comment (jcs090218) [reply README.md]: Added! Thanks! 👍

**Comment by sebastiencs** [comment]
Thanks for working on this @jcs090218 

> I think after-save-hook is fine for now, but @yyoncho suggested using a timer.

I'm not sure if using a timer on `after-change-functions` is the best idea, IMO the `lsp-ui-imenu` window is not really relevant while typing. And this could make quite a lot of requests, with potentially big buffers/slow servers.

What about refreshing it just on `after-save-hook` ? I think it's a more appropriate hook to use, and without the need of a timer

@yyoncho In what use cases a timer would be beneficial ?

**Comment by jcs090218** [comment]
> I'm not sure if using a timer on `after-change-functions` is the best idea, IMO the `lsp-ui-imenu` window is not really relevant while typing. And this could make quite a lot of requests, with potentially big buffers/slow servers.

I think is quite nice to have a updated imenu while typing; of course, this kind of action would take some performance. Have you tried it yet? On my end it work quite well. I have managed to use `run-with-idle-timer` and set to the default delay to `1.0` seconds. It's quite hard to send a refresh request unless you type in a speed of 1 second per letter. Then yes, that's the worst case scenario. In average, I think the user can tweak the `lsp-ui-refresh-delay` base on their own typing habit.

> What about refreshing it just on `after-save-hook` ? I think it's a more appropriate hook to use, and without the need of a timer

I'm fine with this one too. Let me know if I need to revert my updates! Thanks!

**Comment by brotzeit** [comment]
IMO you don't use imenu all the time, but only in certain situations and therefore it's not a big issue that the performance isn't optimal. However it would be nice if imenu would also update the buffer when you are focusing a different buffer than the initial buffer. I just realized that's currently not the case right ?

**Comment by brotzeit** [comment]
@ulises what do you think ?

**Comment by yyoncho** [comment]
> @yyoncho In what use cases a timer would be beneficial ?

this is what other editors do. Also, there is `lsp-on-change-hook` and we do load the symbols info for other components with no perf issues. lsp-treemacs-symbols has optimized rendering since it updates only the changed elements.

**Comment by ulises** [comment]
Apologies for the delay in answering.

My use case would be to have lsp-symbols-imenu open side by side with a buffer that contains a class (or equivalent). So, I'd personally favour 2 hooks:
  - when the main buffer changes (you're working now on a different class or equivalent)
  - when the buffer is saved (you added/removed something in the current class or equivalent)

I'm not sure a timer makes sense considering that workflow, but perhaps I'm just too focused on my own needs :)

**Comment by ulises** [comment]
A further comment, this time on behaviour though: I noticed that when you press `<RET>` on an imenu item the accompanying buffer scrolls to the symbol, but the focus/pointer remains on the imenu buffer. I suspect it'd be nicer if the pointer was placed in the other buffer (perhaps at the beginning of the symbol?).

**Comment by wakira** [comment]
> the focus/pointer remains on the imenu buffer

@ulises Pressing `META-RET` should bring it to the main buffer. Have a try

**Comment by jcs090218** [comment]
> So, I'd personally favour 2 hooks:

I think the current implementation should be good for the 2 use cases you mentioned by binding the `after-change-functions`. :)

**Comment by brotzeit** [comment]
I would give it a try and see if people complain about performance issues.

**Comment by brotzeit** [comment]
Would this be ok for you @sebastiencs ?

**Comment by brotzeit** [comment]
Maybe it would also make sense to allow the user to choose if he wants to use a timer or not.

**Comment by sebastiencs** [comment]
> Would this be ok for you @sebastiencs ?

Alright, let's do it with a timer.
Can we just make sure that the request is made asynchronously ? To not block while waiting for the server.
I don't remember how is made the request but if someone could check if it's already async, if not I think it needs to be implemented @jcs090218

**Comment by jcs090218** [comment]
> Can we just make sure that the request is made asynchronously ? To not block while waiting for the server.
> 
> I don't remember how is made the request but if someone could check if it's already async, if not I think it needs to be implemented @jcs090218 

Any suggestion to make it async? As I have tested; I think the current implementation isn't async? Overall I am not sure if the execution is async/sync because the overall execution goes very fast and I did not feel the delay, yet I assumed is sync.

Right now I have implemented the refresh functionality by reusing the same function `lsp-UI-imenu`. I mean it would be great to have some suggestions from guy who created this function. @sebastiencs Is it you? :)

**Comment by sebastiencs** [comment]
@jcs090218 I just check the code and it's not async.
`lsp-ui-imenu` calls `imenu--make-index-alist` which calls `lsp--imenu-create-index`.
`lsp--imenu-create-index` blocks until it receives a response from the server.

> the overall execution goes very fast and I did not feel the delay

That's true with your specific configuration and server. Some lsp servers might be slow and calling `lsp--imenu-create-index` in a timer will degrade the user experience.
It is not difficult to find on google comments about lsp-ui/mode being slow, so I prefer to be very careful with this kind of thing.

> > Maybe it would also make sense to allow the user to choose if he wants to use a timer or not.

> @brotzeit I think this is a good solution.

Since making it async will require some more work/refactoring, can we disable the timer by default, for this PR ? Until we find a solution to do it asynchronously.
If you want to work on that @jcs090218 I can help you.

> I would give it a try and see if people complain about performance issues.

I don't think it is the best strategy. A few users will complain in the issue tracker, most of them will just stop using lsp-ui.

**Comment by jcs090218** [comment]
> I just check the code and it's not async.

Got you! Thanks for your confirmation! 👍 

> It is not difficult to find on google comments about lsp-ui/mode being slow, so I prefer to be very careful with this kind of thing.

Yes, of course. I agree with you.

> Since making it async will require some more work/refactoring, we can do it in another PR. Can we disable the timer by default, for this PR ? Until we find a solution to do it asynchronously.

Applied! @sebastiencs Let me know if this work! :)

File: lsp-ui-imenu.el
Comment (sebastiencs) [inline lsp-ui-imenu.el:74]: If I understand correctly, this delay is used only when `lsp-ui-imenu-auto-refresh` is non-nil ?
If so, can you rename the variable to `lsp-ui-imenu-auto-refresh-delay`

File: lsp-ui-imenu.el
Comment (sebastiencs) [inline lsp-ui-imenu.el:79]: Please remove the suffix `-p`

File: lsp-ui-imenu.el
Comment (sebastiencs) [inline lsp-ui-imenu.el:70]: Please set this to `nil` by default, as discussed this shouldn't be on by default. Until we make the request async.
And in the docstring please mention that it might slow down navigation/editing because the request is not async.

File: lsp-ui-imenu.el
Comment (sebastiencs) [inline lsp-ui-imenu.el:367]: This shouldn't call `lsp-ui-imenu`, `lsp-ui-imenu` does extra work to create a window.
You will need to split `lsp-ui-imenu`

File: lsp-ui-imenu.el
Comment (sebastiencs) [inline lsp-ui-imenu.el:288]: Only this part of code should be called by `lsp-ui-imenu--refresh`.
See the comment in `lsp-ui-imenu--refresh`

File: lsp-ui-util.el
Comment (sebastiencs) [inline lsp-ui-util.el:35]: Please rename this to `lsp-ui-util--safe-kill-timer`, it's a private function.
And `tmr` to `timer`

File: lsp-ui-imenu.el
Comment (sebastiencs) [inline lsp-ui-imenu.el:95]: This should be a local variable, I think ?

File: lsp-ui-imenu.el
Comment (sebastiencs) [inline lsp-ui-imenu.el:264]: The mode is enabled here but it's not disabled anywhere

**Review by sebastiencs — CHANGES_REQUESTED** [review]

File: lsp-ui-imenu.el
Comment (jcs090218) [reply lsp-ui-imenu.el:95]: There is only one imenu window right? If this is true, the I don't see the benefit setting it to local.

File: lsp-ui-imenu.el
Comment (jcs090218) [reply lsp-ui-imenu.el:70]: This is for `after-save`. I have temporary created another variable `lsp-ui-imenu-refresh-after-change` to disable automatically refresh imenu when `after-change`. So by default; the user can still have auto refresh when `after-save`, but not for the `after-change` because the variable `lsp-ui-imenu-refresh-after-change` is already set to `nil` by default. Let me know if you still want this to be disable by default.

File: lsp-ui-imenu.el
Comment (jcs090218) [reply lsp-ui-imenu.el]: Changed! Thanks!

File: lsp-ui-imenu.el
Comment (jcs090218) [reply lsp-ui-imenu.el]: Fixed! Thanks!

File: lsp-ui-imenu.el
Comment (jcs090218) [reply lsp-ui-imenu.el:264]: Yes. I assumed `imenu` get activated when user calls `M-x lsp-ui-imenu` and not anywhere else. I didn't consider closing it because there is no exit point by the current design? 😕 We might have to discuss this too. Or maybe I have overlooked the code when I'm looking for the exit point.

File: lsp-ui-util.el
Comment (jcs090218) [reply lsp-ui-util.el]: Fixed! Thanks!

File: lsp-ui-imenu.el
Comment (jcs090218) [reply lsp-ui-imenu.el:288]: You meant the code from L265 to L288, right? Aren't the code 99% of the function `lsp-ui-imenu`? Any objection to just call the function? 😕

Comment (jcs090218) [reply lsp-ui-imenu.el:288]: Okay, I may know why. `lsp-ui-menu` created the window and it constantly pop out.

**Comment by brotzeit** [comment]
@jcs090218 can you please bump the version as the last commit.

File: lsp-ui-imenu.el
Comment (jcs090218) [reply lsp-ui-imenu.el]: Okay, I have extracted the refresh execution to `lsp-ui-imenu--refresh-content` function.

**Comment by jcs090218** [comment]
> can you please bump the version as the last commit.

I have bumped version to `6.3`. Does that work for you? Thanks!

File: lsp-ui-imenu.el
Comment (sebastiencs) [reply lsp-ui-imenu.el:70]: hmm what about having just one variable `lsp-ui-imenu-auto-refresh` with 3 values:
- `t` for always
- `after-save` to ignore `after-change-functions`
- `nil` to disable

This will avoid having another variable l`sp-ui-imenu-refresh-after-change`

In all cases, both behaviors should be disabled by default, until we make it async

File: lsp-ui-imenu.el
Comment (sebastiencs) [reply lsp-ui-imenu.el:95]: You can have multiple frames of emacs open, so there can be more than 1 imenu window

File: lsp-ui-imenu.el
Comment (sebastiencs) [reply lsp-ui-imenu.el]: I think it should be disabled when the window of `lsp-ui-imenu` is closed

File: lsp-ui-imenu.el
Comment (jcs090218) [reply lsp-ui-imenu.el]: Got you! Fixed! 👍

File: lsp-ui-imenu.el
Comment (jcs090218) [reply lsp-ui-imenu.el]: See https://github.com/emacs-lsp/lsp-ui/pull/515/commits/1894bc88337576d5553b17449d7c17fc76ad8f37. Does this work for you?

File: lsp-ui-imenu.el
Comment (jcs090218) [reply lsp-ui-imenu.el:70]: > hmm what about having just one variable lsp-ui-imenu-auto-refresh with 3 values:

Changed.

> In all cases, both behaviors should be disabled by default, until we make it async

Got you! I have make the variable default to `nil`.

**Comment by brotzeit** [comment]
> I have bumped version to 6.3. Does that work for you? Thanks!

I mean just before we are merging this pull request, so every commit is included in the new version.

**Comment by jcs090218** [comment]
@brotzeit Got it. Just want to confirm if I did that correctly. 😖

**Comment by jcs090218** [comment]
Okay, I think I have implemented this the way @sebastiencs wants. Let me know if something you want me to change and I will fix it in the later patch! Merge this now!

**Pull request merged**
